docs: small updates for latest pattern compliance (#6227)
snapcore/snapcraft: docs/explanation/snapcraft-yaml.rst

.. meta::
    :description: Explanation of snapcraft.yaml, the project file used by Snapcraft that defines the contents and structure of a snap.

:relatedlinks: [YAML&#32;specification](https://yaml.org)

.. _explanation-snapcraft-yaml:

About snapcraft.yaml
====================

Every snap project depends on a file called ``snapcraft.yaml``, which tells Snapcraft
how to build the snap.

This project file is written in YAML, and declares key-value pairs for all the

> configuration and contents. The :ref:`snapcraft.yaml reference

Reviewer: This one's _very_ optional:
```suggestion
This project file is written in YAML and declares the snap's configuration
and contents with keys. The :ref:`snapcraft.yaml reference
```
When I hear 'key-value pairs', I expect the value to be a primitive, not a nested dict like you'll often find in our project files.

Author: Rewrote this section. f06a056

<reference-snapcraft-yaml>` provides a complete index of the keys and accepted values in
a project file.

> The file can be stored in any of the locations in a project:

Reviewer: ```suggestion
The file can be stored in any of the following locations in a project:
```

Author: Rewrote this section. f06a056

- ``<project>/snapcraft.yaml``
- ``<project>/.snapcraft.yaml``
- ``<project>/snap/snapcraft.yaml``
- ``<project>/build-aux/snap/snapcraft.yaml``

The file is usually organized into four sections:

- Top-level keys
- Platform keys
- App keys
- Part keys, which describe how to import and build the apps inside the snap.

This structure isn't mandated, though. As long as the file is valid YAML and each
key-value pair follows its expected signature, the snap will build.

> For covering the anatomy of ``snapcraft``, we'll examine the snap for `wethr <https://github.com/twobucks/wethr>`__, which is written in Node.

Reviewer: ```suggestion
To cover the anatomy of ``snapcraft.yaml``, we'll examine the snap for `wethr <https://github.com/twobucks/wethr>`__, which is written with Node.
```

Author: Rewrote this section. f06a056

Top-level keys
--------------

The top-level keys provide information about the snap and describe its basic properties

> and information about the software.

Reviewer: ```suggestion
The top-level keys describe the snap's basic properties and its software.
```

Author: f06a056

.. meta::
    :description: Explanation of snapcraft.yaml, the project file used by Snapcraft that defines the contents and structure of a snap.

:relatedlinks: [YAML&#32;specification](https://yaml.org)

.. _explanation-snapcraft-yaml:

About snapcraft.yaml
====================

Every snap project depends on a file called ``snapcraft.yaml``, which tells Snapcraft
how to build the snap.

This project file is written in YAML, and declares key-value pairs for all the
configuration and contents. The :ref:`snapcraft.yaml reference
<reference-snapcraft-yaml>` provides a complete index of the keys and accepted values in
a project file.

The file can be stored in any of the locations in a project:

- ``<project>/snapcraft.yaml``
- ``<project>/.snapcraft.yaml``
- ``<project>/snap/snapcraft.yaml``
- ``<project>/build-aux/snap/snapcraft.yaml``

The file is usually organized into four sections:

- Top-level keys
- Platform keys
- App keys
- Part keys, which describe how to import and build the apps inside the snap.

> This structure isn't mandated, thought. As long as the file is valid YAML and each

Reviewer 2: ```suggestion
This structure isn't mandated, though. As long as the file is valid YAML and each
```

Author: Oof. f299aac

key-value pair follows its expected signature, the snap will build.

We'll examine the anatomy of the ``snapcraft.yaml`` for the `wethr

> <https://github.com/twobucks/wethr>`__ app, written in Node.

Reviewer 2: This was a jarring transition. How about some sort of lead-in, like "to demonstrate..."?

Author: Agreed. f299aac

Top-level keys
--------------

The top-level keys provide information about the snap and describe its basic properties
and information about the software.

.. literalinclude:: code/wethr-snapcraft.yaml
    :caption: wethr project file
    :language: yaml
    :lines: 4-11

.. _reference-anatomy-of-snapcraft-yaml-metadata:
.. _explanation-anatomy-snapcraft-yaml-project-information:

Project information
~~~~~~~~~~~~~~~~~~~

For convenience, the first set of keys comprise information about the snap, such as its
name, version, summary, and description. These identifying keys allow users to find the

> snap through the Snap Store. It also provides some crucial When the snap is published,

Reviewer 2: Looks like you a word?

Author: Don't write late at night. f299aac

this information will be made available to users. It can also be used to pre-populate
certain keys in the snap's Snap Store page during upload.

The wethr snap keeps the most basic information about the project, but doesn't provide
contact information or a link to a donation page, because the original software didn't.

Base
~~~~

As a security default, snaps can't see the host's resources. This follows the principle
of least privilege and prevents conflict with other snaps and increases. However, apps
inside the snap still need some location in the host's storage to act as their root
filesystem. They would also benefit if common libraries were drawn from this root file
system rather than being bundled into each.

The *base* is the minimal set of libraries common to most snaps. The base is itself a
snap that is common to any snap that uses the base. It's mounted and used as the root
filesystem for the apps inside the snap. In essence, this means the snaps behave as
though they were running on a system that matches the base.

> Bases correspond to Ubuntu LTS releases, and are named ``core-<version>``. For example,

Reviewer 2: ```suggestion
Bases correspond to Ubuntu LTS releases, and are named ``core<version>``. For example,
```

Author: Really, don't write late at night. f299aac

the library set in core26 is equivalent to a subset found in Ubuntu 26.04 LTS.

The wethr snap uses core26 as its base.

Quality grade
~~~~~~~~~~~~~

The snap's grade defines the quality level of the snap. Two levels are available,
``devel`` and ``stable``. Snaps with the ``devel`` grade can't be uploaded to either of
the stable or candidate channels in the Snap Store.

The wethr snap has been built and tested over a period of years, so it has a stable
grade.

Confinement
~~~~~~~~~~~

Security confinement distinguishes snaps from most traditional Linux software.
:external+snap:ref:`explanation-security-snap-confinement` provides a high level of
isolation and security, and prevents snaps from being affected by underlying system
changes, snaps affecting each other, or snaps affecting the host.

> The snaps' confinement defines its default access to the host. Confinement is set to one

Reviewer 2: ```suggestion
The snap's confinement defines its default access to the host. Confinement is set to one
```

Author: f299aac

of three levels, higher levels acting as progressively stricter filters:

> - ``strict``. This confinement level uses Linux kernel security features to lock down

Reviewer 2: I think these might look better as deeper-nested headings instead of bullets

Author: This comes from the original file. Instead of new headings, I made them regular paragraphs. f299aac

  the apps inside the snap. By default, a strictly-confined snaps can't access the

>   network, the users' home directory, any audio subsystems or webcams, and it can't

Reviewer 2: ```suggestion
  network, the user's home directory, any audio subsystems or webcams, and it can't
```

Author: f299aac

  display any graphical output through X or Wayland. With strict confinement, if an app

>   needs access to a system resource, it must be granted exhaustively.

Reviewer 2: Did you mean "explicitly" instead of "exhaustively" here?

Author: I mean that the user can't grant an interface to all apps in one go. They must exhaustively specify for each. https://www.merriam-webster.com/dictionary/exhaustively
